### PR TITLE
Cover non-empty-string in `substr`

### PR DESCRIPTION
**Pull request by staabm**
Closes https://github.com/phpstan/phpstan/issues/5324

see https://3v4l.org/aKs2O

### CODE REVIEW

File: tests/PHPStan/Analyser/data/non-empty-string.php [hunk 1]
```diff
@@ -134,6 +134,10 @@ public function doEmpty2(string $s): void
 		}
 	}
 
+	/** @param non-empty-string $nonEmpty */
+	public function doNegativeSubstr($nonEmpty) {
+		assertType[substr($nonEmpty, -5)]
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/non-empty-string.php:139]: Hi, I definitely want to have `substr($nonEmpty, 0, positive-int)` covered as well. Also please investigate what happens with `substr($nonEmpty, 0, negative int)` and if it's worth covering as well.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/data/non-empty-string.php (hunk 1 above)
Comment (staabm) [reply tests/PHPStan/Analyser/data/non-empty-string.php]: Per https://3v4l.org/aKs2O Negative offsets seem not to be relevant for the non-empty-string case

File: src/Type/Php/SubstrDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		$args = $functionCall->args;
+		if (count($args) === 0) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+    if (count($args) >= 2) {
+      $string = $scope->getType($args[0]->value);
+      $offset = $scope->getType($args[1]->value);
+		
+		$negativeOffset = $offset instanceof ConstantIntegerType && $offset->getValue() < 0;
```
Comment (staabm) [inline src/Type/Php/SubstrDynamicReturnTypeExtension.php:38]: Needs to take InterRangeType into account

File: tests/PHPStan/Analyser/data/non-empty-string.php [hunk 3]
```diff
@@ -134,6 +134,15 @@ public function doEmpty2(string $s): void
 		}
 	}
 
+	/**
+	 * @param non-empty-string $nonEmpty
+	 * @param positive-int $positiveInt
+	 */
+	public function doSubstr($nonEmpty, $positiveInt): void {
+		 assertType('non-empty-string', substr($nonEmpty, -5));
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/non-empty-string.php:142]: Needs more test cases for regular string type

**Comment by staabm** [comment]
Except CS this seems good to go. 

Remaing error look unrelated

**Comment by staabm** [comment]
Hmm the error is related.. need to investigate

**Comment by staabm** [comment]
ready for merge

File: src/Type/Php/SubstrDynamicReturnTypeExtension.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		if (count($args) >= 2) {
+			$string = $scope->getType($args[0]->value);
+			$offset = $scope->getType($args[1]->value);
+
+			$negativeOffset = $this->getIntValue($offset) < 0;
+			$zeroOffset = $this->getIntValue($offset) === 0;
+			$positiveLength = false;
```
Comment (ondrejmirtes) [inline src/Type/Php/SubstrDynamicReturnTypeExtension.php:42]: You should ask about whether a number is positive or negative this way:

```php
IntegerRangeType::fromInterval(null, -1)->isSuperTypeOf($offset)->yes() // is negative
IntegerRangeType::fromInterval(1, null)->isSuperTypeOf($offset)->yes() // is positive
```

It also accounts for unions of ConstantIntegerType, like `1|3`.

File: src/Type/Php/SubstrDynamicReturnTypeExtension.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	private function getIntValue(Type $type): ?int
+	{
+		if ($type instanceof IntegerRangeType) {
+			return $type->getMin();
+		}
+		if ($type instanceof ConstantIntegerType) {
+			return $type->getValue();
+		}
+				return null;
```
Comment (ondrejmirtes) [inline src/Type/Php/SubstrDynamicReturnTypeExtension.php:68]: This indent is definitely wrong.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/SubstrDynamicReturnTypeExtension.php (hunk 4 above)
Comment (staabm) [reply src/Type/Php/SubstrDynamicReturnTypeExtension.php:42]: thx, fixed.

Comment (staabm) [reply src/Type/Php/SubstrDynamicReturnTypeExtension.php:42]: does this mean I should check for a zero offset with `IntegerRangeType::fromInterval(0, 0)->isSuperTypeOf($offset)->yes()` ?

Comment (ondrejmirtes) [reply src/Type/Php/SubstrDynamicReturnTypeExtension.php:42]: You can check for that with:

```php
(new ConstantIntegerType(0))->isSuperTypeOf($offset)->yes()
```

Comment (staabm) [reply src/Type/Php/SubstrDynamicReturnTypeExtension.php:41]: > ConstantIntegerType, like `1|3`.

also added testcases to cover that

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**